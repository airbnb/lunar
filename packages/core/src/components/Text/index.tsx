import React from 'react';
import { mutuallyExclusiveTrueProps } from 'airbnb-prop-types';
import withStyles, { css, WithStylesProps } from '../../composers/withStyles';

const sizingProp = mutuallyExclusiveTrueProps('micro', 'small', 'large');
const emphasisProp = mutuallyExclusiveTrueProps('bold', 'light');
const stateProp = mutuallyExclusiveTrueProps('disabled', 'muted', 'inverted');
const alignProp = mutuallyExclusiveTrueProps('centerAlign', 'endAlign', 'startAlign');

export type Props = {
  /** Render the text inline instead of block. */
  baseline?: boolean;
  /** Apply bold emphasis. */
  bold?: boolean;
  /** Align the text to the center. */
  centerAlign?: boolean;
  /** The text to render. */
  children?: React.ReactNode;
  /** Mark the text as disabled. */
  disabled?: boolean;
  /** Align the text to the end. */
  endAlign?: boolean;
  /** Render the text inline-block instead of block. */
  inline?: boolean;
  /** Invert text colors. */
  inverted?: boolean;
  /** Increase font size to large. */
  large?: boolean;
  /** Apply light emphasis. */
  light?: boolean;
  /** Decrease font size to small. */
  micro?: boolean;
  /** Mark the text as muted. */
  muted?: boolean;
  /** Disable text and white space wrapping. */
  noWrap?: boolean;
  /** Preserve whitespace at the beginning and end of children. */
  preserveWhitespace?: boolean;
  /** Decrease font size to small. */
  small?: boolean;
  /** Align the text to the start. */
  startAlign?: boolean;
  /** Truncate the text with an ellipsis. */
  truncated?: boolean;
  /** Uppercase all text. */
  uppercased?: boolean;
};

/** Display a string of text with pre-defined sizing, emphasis, and state styling. */
export class Text extends React.Component<Props & WithStylesProps> {
  static propTypes = {
    bold: emphasisProp,
    centerAlign: alignProp,
    disabled: stateProp,
    endAlign: alignProp,
    inverted: stateProp,
    large: sizingProp,
    light: emphasisProp,
    micro: sizingProp,
    muted: stateProp,
    small: sizingProp,
    startAlign: alignProp,
  };

  static defaultProps = {
    baseline: false,
    bold: false,
    centerAlign: false,
    children: null,
    disabled: false,
    endAlign: false,
    inline: false,
    inverted: false,
    large: false,
    light: false,
    micro: false,
    muted: false,
    noWrap: false,
    preserveWhitespace: false,
    small: false,
    startAlign: false,
    truncated: false,
    uppercased: false,
  };

  render() {
    const {
      baseline,
      bold,
      centerAlign,
      children,
      disabled,
      endAlign,
      inline,
      inverted,
      large,
      light,
      micro,
      muted,
      noWrap,
      preserveWhitespace,
      small,
      startAlign,
      styles,
      truncated,
      uppercased,
    } = this.props;
    let Tag: 'div' | 'span' | 'small' | 'h4' = 'div';

    if (inline) {
      Tag = 'span';
    } else if (micro) {
      Tag = 'small';
    } else if (large) {
      Tag = 'h4';
    }

    return (
      <Tag
        {...css(
          styles.text,
          bold && styles.text_bold,
          disabled && styles.text_disabled,
          inline && styles.text_inline,
          baseline && styles.text_baseline,
          inverted && styles.text_inverted,
          large && styles.text_large,
          light && styles.text_light,
          micro && styles.text_micro,
          muted && styles.text_muted,
          preserveWhitespace && styles.text_preserveWhitespace,
          small && styles.text_small,
          truncated && styles.text_truncated,
          uppercased && styles.text_uppercased,
          centerAlign && styles.text_center,
          endAlign && styles.text_end,
          startAlign && styles.text_start,
          noWrap && styles.text_noWrap,
        )}
      >
        {children}
      </Tag>
    );
  }
}

export default withStyles(({ color, font, pattern }) => ({
  text: {
    ...font.textReset,
    ...font.textRegular,
    color: color.accent.text,
  },

  text_baseline: {
    display: 'inline',
  },

  text_inline: {
    display: 'inline-block',
  },

  text_preserveWhitespace: {
    whiteSpace: 'pre-wrap',
  },

  text_micro: {
    ...font.textMicro,
  },

  text_small: {
    ...font.textSmall,
  },

  text_large: {
    ...font.textLarge,
  },

  text_disabled: {
    ...pattern.disabled,
  },

  text_inverted: {
    color: color.base,
  },

  text_muted: {
    color: color.muted,
  },

  text_bold: {
    fontWeight: font.weights.semibold,
  },

  text_light: {
    fontWeight: font.weights.light,
  },

  text_truncated: {
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',

    '@selectors': {
      '> div': {
        display: 'inline',
      },
    },
  },

  text_uppercased: {
    textTransform: 'uppercase',
  },

  text_center: {
    textAlign: 'center',
  },

  text_end: {
    textAlign: 'right',
  },

  text_start: {
    textAlign: 'left',
  },

  text_noWrap: {
    whiteSpace: 'nowrap',
  },
}))(Text);
