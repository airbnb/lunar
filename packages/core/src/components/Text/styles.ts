import { StyleSheet } from '../../hooks/useStyles';

const styleSheet: StyleSheet = ({ color, font, pattern }) => ({
  text: {
    ...font.textReset,
    ...font.textRegular,
    color: color.accent.text,
  },

  text_baseline: {
    display: 'inline',
  },

  text_inline: {
    display: 'inline-block',
  },

  text_preserveWhitespace: {
    whiteSpace: 'pre-wrap',
  },

  text_micro: {
    ...font.textMicro,
  },

  text_small: {
    ...font.textSmall,
  },

  text_large: {
    ...font.textLarge,
  },

  text_disabled: {
    ...pattern.disabled,
  },

  text_inverted: {
    color: color.base,
  },

  text_muted: {
    color: color.muted,
  },

  text_bold: {
    fontWeight: font.weights.semibold,
  },

  text_light: {
    fontWeight: font.weights.light,
  },

  text_truncated: {
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',

    '@selectors': {
      '> div': {
        display: 'inline',
      },
    },
  },

  text_uppercased: {
    textTransform: 'uppercase',
  },

  text_uppercased_micro: {
    letterSpacing: 1,
  },

  text_center: {
    textAlign: 'center',
  },

  text_end: {
    textAlign: 'right',
  },

  text_start: {
    textAlign: 'left',
  },

  text_noWrap: {
    whiteSpace: 'nowrap',
  },
});

export default styleSheet;

export { styleSheet };
