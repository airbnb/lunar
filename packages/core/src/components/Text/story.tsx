import React from 'react';
import LoremIpsum from ':storybook/components/LoremIpsum';
import Text from '.';

export default {
  title: 'Core/Text',
  parameters: {
    inspectComponents: [Text],
  },
};

export function aBasicStringOfText() {
  return (
    <Text>
      <LoremIpsum />
    </Text>
  );
}

aBasicStringOfText.story = {
  name: 'A basic string of text.',
};

export function withLightBoldAndUppercasedEmphasis() {
  return (
    <>
      <Text light>
        <LoremIpsum />
      </Text>
      <br />
      <Text bold>
        <LoremIpsum />
      </Text>
      <br />
      <Text uppercased>
        <LoremIpsum />
      </Text>
    </>
  );
}

withLightBoldAndUppercasedEmphasis.story = {
  name: 'With light, bold, and uppercased emphasis.',
};

export function withDifferentSizingMicroSmallRegularDefaultAndLarge() {
  return (
    <>
      <Text micro>
        <LoremIpsum />
      </Text>
      <br />
      <Text small>
        <LoremIpsum />
      </Text>
      <br />
      <Text>
        <LoremIpsum />
      </Text>
      <br />
      <Text large>
        <LoremIpsum />
      </Text>
    </>
  );
}

withDifferentSizingMicroSmallRegularDefaultAndLarge.story = {
  name: 'With different sizing: micro, small, regular (default), and large.',
};

export function withDifferentStatesMutedDisabledAndInverted() {
  return (
    <>
      <Text muted>
        <LoremIpsum short />
      </Text>
      <Text disabled>
        <LoremIpsum short />
      </Text>
      <Text inverted>
        <LoremIpsum short />
      </Text>
    </>
  );
}

withDifferentStatesMutedDisabledAndInverted.story = {
  name: 'With different states: muted, disabled, and inverted.',
};

export function withWhitespacePreserved() {
  return (
    <Text preserveWhitespace>
      {'     '}
      <LoremIpsum short />
      {'     '}
    </Text>
  );
}

withWhitespacePreserved.story = {
  name: 'With whitespace preserved:',
};

export function withTruncated() {
  return (
    <Text truncated>
      <LoremIpsum short /> <LoremIpsum />
    </Text>
  );
}

withTruncated.story = {
  name: 'With truncated.',
};

export function withAlignedText() {
  return (
    <div style={{ textAlign: 'center' }}>
      <Text>
        <Text bold>Parent alignment</Text> <LoremIpsum short />
      </Text>
      <br />
      <Text startAlign>
        <Text bold>Start align</Text> <LoremIpsum short />
      </Text>
      <br />
      <Text centerAlign>
        <Text bold>Center align</Text> <LoremIpsum short />
      </Text>
      <br />
      <Text endAlign>
        <Text bold>End align</Text> <LoremIpsum short />
      </Text>
    </div>
  );
}

withAlignedText.story = {
  name: 'With aligned text.',
};
