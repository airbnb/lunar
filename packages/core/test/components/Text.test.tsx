import React from 'react';
import { shallowWithStyles } from '@airbnb/lunar-test-utils';
import Text from '../../src/components/Text';

describe('<Text />', () => {
  it('errors when multiple alignments are used at once', () => {
    expect(() => {
      shallowWithStyles(
        <Text startAlign endAlign>
          Default
        </Text>,
      );
    }).toThrowError();
  });

  it('errors when multiple states are used at once', () => {
    expect(() => {
      shallowWithStyles(
        <Text muted inverted>
          Default
        </Text>,
      );
    }).toThrowError();
  });

  it('errors when multiple emphasis are used at once', () => {
    expect(() => {
      shallowWithStyles(
        <Text light bold>
          Default
        </Text>,
      );
    }).toThrowError();
  });

  it('errors when multiple sizes are used at once', () => {
    expect(() => {
      shallowWithStyles(
        <Text small large>
          Default
        </Text>,
      );
    }).toThrowError();
  });

  it('can render without children', () => {
    const wrapper = shallowWithStyles(<Text />);

    expect(wrapper).toMatchSnapshot();
  });

  it('uses `div` tag by default', () => {
    const wrapper = shallowWithStyles(<Text>Default</Text>);

    expect(wrapper.type()).toBe('div');
  });

  it('changes tag to `span` if `inline` is passed', () => {
    const wrapper = shallowWithStyles(<Text inline>Default</Text>);

    expect(wrapper.type()).toBe('span');
  });

  it('changes tag to `small` if `micro` is passed', () => {
    const wrapper = shallowWithStyles(<Text micro>Default</Text>);

    expect(wrapper.type()).toBe('small');
  });

  it('changes tag to `h4` if `large` is passed', () => {
    const wrapper = shallowWithStyles(<Text large>Default</Text>);

    expect(wrapper.type()).toBe('h4');
  });

  it('renders light', () => {
    const wrapper = shallowWithStyles(<Text light>Light</Text>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders bold', () => {
    const wrapper = shallowWithStyles(<Text bold>Bold</Text>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders micro', () => {
    const wrapper = shallowWithStyles(<Text micro>Micro</Text>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders small', () => {
    const wrapper = shallowWithStyles(<Text small>Small</Text>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders large', () => {
    const wrapper = shallowWithStyles(<Text large>Large</Text>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders disabled', () => {
    const wrapper = shallowWithStyles(<Text disabled>Disabled</Text>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders inline', () => {
    const wrapper = shallowWithStyles(<Text inline>Inline</Text>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders inverted', () => {
    const wrapper = shallowWithStyles(<Text inverted>Inverted</Text>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders muted', () => {
    const wrapper = shallowWithStyles(<Text muted>Muted</Text>);

    expect(wrapper).toMatchSnapshot();
  });

  it('renders with preserved whitespace', () => {
    const wrapper = shallowWithStyles(
      <Text preserveWhitespace>
        {'   '}
        Preserved whitespace
        {'   '}
      </Text>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders uppercased', () => {
    const wrapper = shallowWithStyles(<Text uppercased>Uppercased</Text>);

    expect(wrapper).toMatchSnapshot();
  });
});
