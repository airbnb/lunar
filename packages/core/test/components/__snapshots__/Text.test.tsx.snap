// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Text /> can render without children 1`] = `
<div
  className="text_8h9m6q"
/>
`;

exports[`<Text /> renders bold 1`] = `
<div
  className="text_8h9m6q-o_O-text_bold_pgfqnw"
>
  Bold
</div>
`;

exports[`<Text /> renders disabled 1`] = `
<div
  className="text_8h9m6q-o_O-text_disabled_5k6q5e"
>
  Disabled
</div>
`;

exports[`<Text /> renders inline 1`] = `
<span
  className="text_8h9m6q-o_O-text_inline_36rlri"
>
  Inline
</span>
`;

exports[`<Text /> renders inverted 1`] = `
<div
  className="text_8h9m6q-o_O-text_inverted_bd4m75"
>
  Inverted
</div>
`;

exports[`<Text /> renders large 1`] = `
<h4
  className="text_8h9m6q-o_O-text_large_1aiv2f"
>
  Large
</h4>
`;

exports[`<Text /> renders light 1`] = `
<div
  className="text_8h9m6q-o_O-text_light_n2wlx5"
>
  Light
</div>
`;

exports[`<Text /> renders micro 1`] = `
<small
  className="text_8h9m6q-o_O-text_micro_azj9az"
>
  Micro
</small>
`;

exports[`<Text /> renders muted 1`] = `
<div
  className="text_8h9m6q-o_O-text_muted_10k87om"
>
  Muted
</div>
`;

exports[`<Text /> renders small 1`] = `
<div
  className="text_8h9m6q-o_O-text_small_krq4jw"
>
  Small
</div>
`;

exports[`<Text /> renders uppercased 1`] = `
<div
  className="text_8h9m6q-o_O-text_uppercased_4efw5a"
>
  Uppercased
</div>
`;

exports[`<Text /> renders with preserved whitespace 1`] = `
<div
  className="text_8h9m6q-o_O-text_preserveWhitespace_11oyobo"
>
     
  Preserved whitespace
     
</div>
`;
